pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -83,16 +85,37 @@ func (s *FlinkStateMachine) shouldRollback(ctx context.Context, application *v1a
 		return false
 	}
 
-	appLastUpdated := application.Status.LastUpdatedAt
-	if appLastUpdated != nil && !v1alpha1.IsRunningPhase(application.Status.Phase) {
-		elapsedTime := s.clock.Since(appLastUpdated.Time)
-		if s.getStalenessDuration() > 0 && elapsedTime > s.getStalenessDuration() {
+	if application.Status.LastSeenError == nil {
+		return false
+	}
+	// Check if the error is retryable
+	if s.retryHandler.IsErrorRetryable(application.Status.LastSeenError) {
+		if s.retryHandler.IsRetryRemaining(application.Status.LastSeenError, application.Status.RetryCount) {
+			s.flinkController.LogEvent(ctx, application, "", corev1.EventTypeWarning, fmt.Sprintf("Application in phase %v retrying with error %v", application.Status.Phase, application.Status.LastSeenError))
+			return false
+		}
+		// Retryable error with retries exhausted
+		application.Status.LastSeenError = nil
+		return true
+	}
+
+	// For a class of errors that should fail fast, e.g. submitjob failures, always fail fast
+	if s.retryHandler.IsErrorFailFast(application.Status.LastSeenError) {
+		s.flinkController.LogEvent(ctx, application, "", corev1.EventTypeWarning, fmt.Sprintf("Application failed to progress in the %v phase with error %v", application.Status.Phase, application.Status.LastSeenError))
+		application.Status.LastSeenError = nil
+		return true
+	}
+
+	// All other errors, use a time based wait to determine whether to rollback or not.
+	if application.Status.LastSeenError != nil && application.Status.LastUpdatedAt != nil {
```

Comment:
I don't think this logic quite works for issues that arise during cluster start-up (for example, wrong image). The lastSeenError is never set (as the error occurs in the pod as is not visible to the operator).

Reply:
Good point. Updated to have a generic time-based fallback for cases (with or without LastSeenErrors) that don't fall under retryable/fail fast categories.  Side note: Pod/cluster based errors are really hard to debug right now :(